rust/agents/relayer/src/msg/processor.rs

```diff
@@ -0,0 +1,239 @@
+            // the DB is now coherent.
+            // TODO(webbhorn): Why can't we yield here instead of sleep?
+            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            return Ok(());
+        };
+
+        // Skip if for different inbox.
+        if message.message.destination != self.inbox_contracts.inbox.local_domain() {
+            info!(
+                inbox_name=?self.inbox_contracts.inbox.chain_name(),
+                local_domain=?self.inbox_contracts.inbox.local_domain(),
+                dst=?message.message.destination,
+                msg=?message,
+                "message not for local domain, skipping idx {}", self.message_leaf_index);
+            self.message_leaf_index += 1;
+            return Ok(());
+        }
+
+        // Skip if not whitelisted.
+        if !self.whitelist.msg_matches(&message.message) {
+            info!(
+                inbox_name=?self.inbox_contracts.inbox.chain_name(),
+                local_domain=?self.inbox_contracts.inbox.local_domain(),
+                dst=?message.message.destination,
+                whitelist=?self.whitelist,
+                msg=?message,
+                "message not whitelisted, skipping idx {}", self.message_leaf_index);
+            self.message_leaf_index += 1;
+            return Ok(());
+        }
+
+        // If validator hasn't published checkpoint covering self.message_leaf_index yet, wait
+        // until it has, before forwarding the message to the submitter channel.
+        let mut ckpt;
+        loop {
+            ckpt = self.ckpt_rx.borrow().clone();
+            match &ckpt {
+                Some(ckpt) if ckpt.checkpoint.index >= self.message_leaf_index => {
+                    break;
+                }
+                _ => {
+                    self.ckpt_rx.changed().await?;
+                }
+            }
+        }
+        let checkpoint = ckpt.unwrap();
+        assert!(checkpoint.checkpoint.index >= self.message_leaf_index);
+
+        // Include proof against checkpoint for message in the args provided to the submitter.
+        if checkpoint.checkpoint.index >= self.prover_sync.count() {
+            self.prover_sync
+                .update_to_checkpoint(&checkpoint.checkpoint)
+                .await?;
+        }
+        assert_eq!(checkpoint.checkpoint.index + 1, self.prover_sync.count());
+        let proof = self.prover_sync.get_proof(self.message_leaf_index)?;
+
+        // Finally, build the submit arg and dispatch it to the submitter.
+        let submit_args = SubmitMessageArgs {
+            leaf_index: self.message_leaf_index,
+            committed_message: message,
+            checkpoint,
+            proof,
+            num_retries: 0,
+        };
+        if self
+            .db
+            .leaf_by_leaf_index(self.message_leaf_index)?
+            .is_some()
+        {
+            info!(
+                "sending message at idx {} to submitter",
```

Comment:
Wondering if this could be more clear to someone who's just operating the relayer but not familiar with the code